# Análisis de ventas

Se facilita una lista de `dict` con datos reales de ventas con las siguientes keys:

* 'Order ID'
* 'Product'
* 'Quantity Ordered'
* 'Price Each
* 'Order Date'
* 'Purchase Address'

y values de tipo `str` en todos los casos:

In [1]:
import pickle

with open('sales_data.pickle', 'rb') as handle:
    sales_data = pickle.load(handle)

In [2]:
print(len(sales_data))  # tiene 186850 entradas
sales_data[0:1]  # los diccionarios tienen la siguiente estructura

186850


[{'Order ID': '209921',
  'Product': 'USB-C Charging Cable',
  'Quantity Ordered': '1',
  'Price Each': '11.95',
  'Order Date': '06/23/19 19:34',
  'Purchase Address': '950 Walnut St, Portland, ME 04101'}]

In [3]:
sales_keys = sales_data[0].keys()

Responder los siguientes interrogantes de interés, para lo que se requiere:

* Inspeccionar los datos
* Formatearlos adecuadamente
* Elegir y confeccionar nuevas estructuras de datos

In [4]:
sales_data[158]
# encontramos que algunos diccionarios tenian las values iguales a sus keys, por los que los eliminamos para evitar trabajar sobre ellos.

{'Order ID': 'Order ID',
 'Product': 'Product',
 'Quantity Ordered': 'Quantity Ordered',
 'Price Each': 'Price Each',
 'Order Date': 'Order Date',
 'Purchase Address': 'Purchase Address'}

In [5]:
for sale in sales_data:
    for k in sales_keys:
        if sale[k] == k:
            sales_data.pop(sales_data.index(sale))
            break
print(len(sales_data))

186495


In [6]:
# encontramos que algunos diccionarios tenian las values vacias, por los que los eliminamos para evitar trabajar sobre ellos.

for sale in sales_data:
    for k in sales_keys:
        if type(sale[k]) != type('string'):
            sales_data.pop(sales_data.index(sale))
            break
        elif sale[k] == '':
            sales_data.pop(sales_data.index(sale))
            break
len(sales_data)

185950

In [ ]:
## para recorrer la lista de diccionarios se decició recurrir a listas que contengan la información requerida para cada ejercicio de manera ordenada

**1)** ¿Cuál es el mes que reporta **la mayor cantidad de productos**? 

*Se espera*: `El mes de mayor venta es: {Enero/Febrero/Marzo/Abril...}` *(nótese que se debe printear el nombre del mes)*

In [9]:
months_sales_data = []

for sale in sales_data:
    months_sales_data.append(sale['Order Date'][0:2])  #order date es del formato mm/dd/yyyy, hh:mm:ss solo me interesa el mes
months_sales_data[-1]

'07'

In [10]:
months_dict = {}

for i in range(1, 13):
    if i < 10:
        k = '0' + str(i)
    else:
        k = str(i)
    months_dict[k] = 0
# hacer una función para crear diccionario de meses

In [11]:
for month in months_dict.keys():
    for order in months_sales_data:
        if month == order:
            months_dict[month] = months_dict[month] + 1
months_dict    

{'01': 9709,
 '02': 11975,
 '03': 15153,
 '04': 18279,
 '05': 16566,
 '06': 13554,
 '07': 14293,
 '08': 11961,
 '09': 11621,
 '10': 20282,
 '11': 17573,
 '12': 24984}

In [12]:
spanish_months = ['enero', 'febrero', 'marzo', 'abril', 'mayo', 'junio', 'julio', 'agosto', 'setiembre', 'octubre', 'noviembre', 'diciembre'] 
spanish_months_dict = {}

for i in range(1, 13):
    if i < 10:
        k = '0' + str(i)
    else:
        k = str(i)
    spanish_months_dict[k] = spanish_months[i - 1]

In [13]:
sales_comparative = []

for month in months_dict:
    sales_comparative.append(months_dict[month])

sales = max(sales_comparative)
sales

24984

In [14]:
for month in months_dict:
    if str(sales) in str(months_dict[month]):
        best_month = spanish_months_dict[month]
best_month

'diciembre'

In [15]:
print(f'El mes de mayor venta es: {best_month}.')

El mes de mayor venta es: diciembre.


El motivo por el cual las ventas aumentan en diciembre es la celebración de la navidad y su tradición de entregar regalos. 

**2)** ¿Qué ganancia reporta dicho mes?  
*Se espera*: `En el mes de {Enero/Febrero/Marzo/Abril...} se reporta una ganancia total de USD{xxxx.xx}` *(nótese que el monto debe incluir DOS decimales)*

In [16]:
def multiply(a: int, b: float) -> float:
    return a * b

In [17]:
months_gain_data = []

for sale in sales_data:       
    months_gain_data.append( [ sale['Order Date'][0:2] , round(multiply(int(sale['Quantity Ordered']), float(sale['Price Each'])), 2) ] )
# formato: ['07', 2.99]

In [18]:
gains_dict = {}

for i in range(1, 13):
    if i < 10:
        k = '0' + str(i)
    else:
        k = str(i)
    gains_dict[k] = 0

In [26]:
for data in months_gain_data:
    for month in gains_dict.keys():
        if month == data[0]:
            gains_dict[month] += round(data[1],2)
            round(gains_dict[month],2)

print(gains_dict)

{'01': 3644513.460000878, '02': 4404044.840001397, '03': 5614200.760002203, '04': 6781340.480002939, '05': 6305213.500002601, '06': 5155604.520001792, '07': 5295551.520001975, '08': 4488935.760001408, '09': 4195120.260001261, '10': 7473453.760003513, '11': 6399206.400002738, '12': 9226886.680003785}


In [20]:
for month in spanish_months_dict:
    if spanish_months_dict[month] == best_month:
        best_month_gains = round(gains_dict[month],2)
best_month_gains   

4613443.34

In [21]:
print(f'En el mes de {best_month} se reporta una ganancia total de USD {best_month_gains}.')

En el mes de diciembre se reporta una ganancia total de USD 4613443.34.


In [29]:
gains_list = []
for month in gains_dict:
    gains_list.append(gains_dict[month])
total_gains = sum(gains_list)
print(round(total_gains,2))

68984071.94


In [31]:
sales_percentage = (best_month_gains / total_gains)*100
sales_percentage

6.687693564988225

A pesar de que diciembre es el mes con más ventas registradas, no es el mes con más ganancias. Solo representan el 6,7 % de las ganancias totales. Dado que este mes es el de mayor ventas, podría ofrecerse descuentos especiales para incentivar a los clientes a comprar productos más caros y generar así mayores ganancias.

**3)** ¿Qué porcentaje del total de las ventas representa?  
*Se espera*: `Las ventas en el mes de {Enero/Febrero/Marzo/Abril...} representan un xx.xx% del total de las ventas` *(nótese que el monto debe incluir DOS decimales)*

In [32]:
total_sales = sum(sales_comparative)
total_sales

185950

In [33]:
sales_percentage = round((max(sales_comparative)/total_sales) * 100 ,2)
sales_percentage

13.44

In [34]:
print(f'Las ventas en el mes de {best_month} representan un {sales_percentage} % del total de las ventas.')

Las ventas en el mes de diciembre representan un 13.44 % del total de las ventas.


**4)** ¿Qué ciudad reportó la mayor cantidad de ventas?  
*Se espera*: `La ciudad con mayor cantidad de ventas es {Dallas/Los Angeles/Boston...} reporta la mayor cantidad de ventas` *(nótese que se debe especificar la ciudad y no la dirección completa)*

In [35]:
city_sales_data = []

for sale in sales_data:
    city_sales_data.append( sale['Purchase Address'].split(', ')[1])
#print(city_sales_data[-1])

In [36]:
citys_set = set(city_sales_data)  # me arme un set para sacar las keys de mi diccionario de ciudades
citys_sales_dict = {}

for city in citys_set:
    citys_sales_dict[city] = 0
citys_sales_dict

{'Dallas': 0,
 'Boston': 0,
 'San Francisco': 0,
 'Austin': 0,
 'Portland': 0,
 'New York City': 0,
 'Atlanta': 0,
 'Los Angeles': 0,
 'Seattle': 0}

In [37]:
for city in citys_sales_dict.keys():
    for order in city_sales_data:
        if city == order:
            citys_sales_dict[city] = citys_sales_dict[city] + 1
citys_sales_dict   

{'Dallas': 14820,
 'Boston': 19934,
 'San Francisco': 44732,
 'Austin': 9905,
 'Portland': 12465,
 'New York City': 24876,
 'Atlanta': 14881,
 'Los Angeles': 29605,
 'Seattle': 14732}

In [38]:
city_comparative = []
for city in citys_sales_dict:
    city_comparative.append(citys_sales_dict[city])

orders = max(city_comparative)
orders

44732

In [39]:
for city in citys_sales_dict:
    if str(orders) in str(citys_sales_dict[city]):
        best_city = city
best_city

'San Francisco'

In [40]:
print(f'Con {orders} ventas, la ciudad con mayor cantidad de ventas es {best_city}.')

Con 44732 ventas, la ciudad con mayor cantidad de ventas es San Francisco.


San Francisco tiene 874.784 habitantes, mientras que Nueva York tiene 8,38 millones. Sería interesante aumentar la publicidad en esa ciudad para aumentar así también las ventas totales. 

**5)** ¿Qué porcentaje del total de las ventas representa?  
*Se espera*: `Las ventas en {Dallas/Los Angeles/Boston...} representan un xx.xx% del total` *(nótese que se debe especificar la ciudad y no la dirección completa)*

In [41]:
city_percentage = round((max(city_comparative)/total_sales) * 100 ,2)
city_percentage

24.06

In [42]:
print(f'Las ventas en {best_city} representan un {city_percentage} % del total.')

Las ventas en San Francisco representan un 24.06 % del total.


Las ventas en San Francisco representan practicamente 1/4 de las ventas totales de la empresa, marcando una fuerte presencia en la costa oeste.

**6)** ¿En qué horas del día se deberían publicar anuncios promocionales para maximizar las ventas?  
*Se espera*: `Las horas del día en las que más ventas se registran son: {xx}hs y {yy}hs` *(nótese que se requieren horas del día, como "10hs" o "20hs")*

In [44]:
hours_sales_data = []

for sale in sales_data:
    hours_sales_data.append(sale['Order Date'][-5:-3])
hours_sales_data[0]

'19'

In [45]:
hours_dict = {}

for i in range(0, 24):
    if i < 10:
        k = '0' + str(i)
    else:
        k = str(i)
    hours_dict[k] = 0
#hours_dict

In [47]:
for hour in hours_dict.keys():
    for order in hours_sales_data:
        if hour == order:
            hours_dict[hour] = hours_dict[hour] + 1
hours_dict  

{'00': 3910,
 '01': 2350,
 '02': 1243,
 '03': 831,
 '04': 854,
 '05': 1321,
 '06': 2482,
 '07': 4011,
 '08': 6256,
 '09': 8748,
 '10': 10944,
 '11': 12411,
 '12': 12587,
 '13': 12129,
 '14': 10984,
 '15': 10175,
 '16': 10384,
 '17': 10899,
 '18': 12280,
 '19': 12905,
 '20': 12228,
 '21': 10921,
 '22': 8822,
 '23': 6275}

In [48]:
hours_comparative = []

for hour in hours_dict.keys():
    hours_comparative.append(hours_dict[hour])
    
best_hour = max(hours_comparative)
hours_comparative.pop(hours_comparative.index(max(hours_comparative)))
print(hours_comparative)

[3910, 2350, 1243, 831, 854, 1321, 2482, 4011, 6256, 8748, 10944, 12411, 12587, 12129, 10984, 10175, 10384, 10899, 12280, 12228, 10921, 8822, 6275]


In [49]:
second_best_hour = max(hours_comparative)
sales_hours = [best_hour , second_best_hour]
sales_hours

[12905, 12587]

In [50]:
for hour in hours_dict:
    if str(best_hour) in str(hours_dict[hour]):
        best_hour = hour
    elif str(second_best_hour) in str(hours_dict[hour]):
        second_best_hour = hour
print(best_hour)  # después del trabajo, apenas la gente llega a casa
print(second_best_hour)  # en el almuerzo

19
13


In [52]:
print(f'Las horas del día en las que más ventas se registran son: {best_hour}hs y {second_best_hour}hs.') 

Las horas del día en las que más ventas se registran son: 19hs y 13hs.


La mayoría de las ventas tiene lugar durante la hora del almuerzo y luego del horario de oficina, cuando las personas regresan a sus hogares. Para potenciar estas ventas hay que aumentar la publicidad cercana a estas horas, para atraer a los clientes a realizar la compra.

**7)** Determinar productos que se venden juntos y DE A DOS.  
*Se espera*: `Los siguientes productos se venden juntos y de a dos: {x}-{y}, {yy}-{zz} ...`  
**OBS**: Para una misma *orden* se pueden regitrar varios productos.

In [53]:
product_sales_data = []

for sale in sales_data:       
    product_sales_data.append( [sale['Order ID'], [sale['Product']], (int(sale['Quantity Ordered']) ) ])
# formato: ['209921', 'USB-C Charging Cable', 1]
#product_sales_data[0]

In [56]:
order_id_list = []

for sale in sales_data:
    order_id_list.append(sale['Order ID'])
order_id_set = set(order_id_list)
len(order_id_set)

178437

In [57]:
len(product_sales_data)-len(order_id_set)

7513

Hay 7513 order ID que se repiten al menos una vez

In [58]:
order_id_dict = {}

for i_d in order_id_set:
    order_id_dict[i_d] = []
type(order_id_dict[i_d])

list

In [ ]:
for i_d in order_id_dict.keys():
    for data in product_sales_data:
        products = order_id_dict[i_d] + data[1]
        order_id_dict[i_d] = products

In [ ]:
order_id_dict

In [ ]:
two_products_dict ={}
for i_d in order_id_dict.keys():  # para quedarme solo con las compras de 2 productos distintos
    if len(order_id_dict[i_d]) != 2:
        two_products_dict[id] = 0
        

In [ ]:
for i_d in two_products_dict.keys():
    for data in product_sales_data:
        two_products_dict[id] =+ data[2] 

In [ ]:
###### terminar después

In [ ]:
print(f'Los siguientes productos se venden juntos y de a dos: {x}-{y}, {yy}-{zz} ...')

**8)** ¿Cuál es el producto más vendido?
*Se espera*: `El producto más vendido es {x}`

In [54]:
products_list = []

for sale in sales_data:       
     products_list.append(sale['Product'])
# formato: ['209921', 'USB-C Charging Cable', 1]
products_list[0]

'USB-C Charging Cable'

In [55]:
products_set = set(products_list)
products_set

{'20in Monitor',
 '27in 4K Gaming Monitor',
 '27in FHD Monitor',
 '34in Ultrawide Monitor',
 'AA Batteries (4-pack)',
 'AAA Batteries (4-pack)',
 'Apple Airpods Headphones',
 'Bose SoundSport Headphones',
 'Flatscreen TV',
 'Google Phone',
 'LG Dryer',
 'LG Washing Machine',
 'Lightning Charging Cable',
 'Macbook Pro Laptop',
 'ThinkPad Laptop',
 'USB-C Charging Cable',
 'Vareebadd Phone',
 'Wired Headphones',
 'iPhone'}

In [61]:
products_dict = {}
for product in products_set:
    products_dict[product] = 0
#products_dict

In [63]:
for product in products_dict.keys():
    products_dict[product] = products_list.count(product)
products_dict

{'USB-C Charging Cable': 21903,
 'iPhone': 6842,
 'LG Dryer': 646,
 'Bose SoundSport Headphones': 13325,
 '20in Monitor': 4101,
 'Wired Headphones': 18882,
 'LG Washing Machine': 666,
 '27in 4K Gaming Monitor': 6230,
 'Vareebadd Phone': 2065,
 'Flatscreen TV': 4800,
 '34in Ultrawide Monitor': 6181,
 '27in FHD Monitor': 7507,
 'Google Phone': 5525,
 'Macbook Pro Laptop': 4724,
 'AA Batteries (4-pack)': 20577,
 'Lightning Charging Cable': 21658,
 'AAA Batteries (4-pack)': 20641,
 'ThinkPad Laptop': 4128,
 'Apple Airpods Headphones': 15549}

In [64]:
product_amount = []

for product in products_dict.keys():
    product_amount.append(products_dict[product])

amount_most_sold_product = max(product_amount)
amount_most_sold_product    

21903

In [65]:
for product in products_dict.keys():
    if amount_most_sold_product == products_dict[product]:
        most_sold_product = product
most_sold_product

'USB-C Charging Cable'

In [66]:
print(f'El producto más vendido es {most_sold_product}.')

El producto más vendido es USB-C Charging Cable.


**9)** ¿Cuál es la relación entre el precio del producto más vendido y el del menos vendido?
*Se espera:* `El precio del producto más vendido es {n} veces mayor/menor que el producto menos vendido`

In [67]:
amount_worst_sold_product = min(product_amount)
amount_worst_sold_product    

646

In [68]:
for product in products_dict.keys():
    if amount_worst_sold_product == products_dict[product]:
        worst_sold_product = product
worst_sold_product

'LG Dryer'

In [70]:
products_price_dict = {}

for product in products_set:
    products_price_dict[product] = 0
    
#products_price_dict

In [71]:
products_price_dict = {}

for sale in sales_data:
    for product in products_dict.keys():
        if product == sale['Product']:
            products_price_dict[product] = round(float(sale['Price Each']),2)
products_price_dict

{'USB-C Charging Cable': 11.95,
 'Macbook Pro Laptop': 1700.0,
 'ThinkPad Laptop': 999.99,
 '27in FHD Monitor': 149.99,
 'Bose SoundSport Headphones': 99.99,
 'Apple Airpods Headphones': 150.0,
 'Lightning Charging Cable': 14.95,
 'Wired Headphones': 11.99,
 'Flatscreen TV': 300.0,
 'AA Batteries (4-pack)': 3.84,
 'AAA Batteries (4-pack)': 2.99,
 '34in Ultrawide Monitor': 379.99,
 'iPhone': 700.0,
 'Google Phone': 600.0,
 '27in 4K Gaming Monitor': 389.99,
 '20in Monitor': 109.99,
 'LG Dryer': 600.0,
 'Vareebadd Phone': 400.0,
 'LG Washing Machine': 600.0}

In [72]:
for product in products_price_dict.keys():
    if worst_sold_product == product:
        worst_sold_product_price = products_price_dict[product]
    elif most_sold_product == product:
        most_sold_product_price = products_price_dict[product]
print(most_sold_product_price)
print(worst_sold_product_price)

11.95
600.0


In [73]:
n= int(worst_sold_product_price / most_sold_product_price)
n

50

In [74]:
print(f'El precio del producto más vendido es {n} veces menor que el producto menos vendido.')

El precio del producto más vendido es 50 veces menor que el producto menos vendido.


**10)** ¿Cuál es la órden con mayor cantidad de productos?
*Se espera*: `La órden con mayor cantidad de productos es la número {x}, con {y} productos en total`

In [75]:
order_id_product_quantity_dict = {}

for i_d in order_id_set:
    order_id_product_quantity_dict[i_d] = 0

In [ ]:
for i_d in order_id_product_quantity_dict.keys():
    for data in product_sales_data:
        products =+ data[2]
        order_id_product_quantity_dict[i_d] = products

In [ ]:
product_quantity = []

for i_d in order_id_product_quantity_dict.keys():
    order_id_product_quantity_dict[i_d] = len(order_id_dict[i_d])
most_products_in_one_order = max(product_quantity)

In [ ]:
for i_d in order_id_product_quantity_dict.keys():
    if most_products_in_one_order in order_id_product_quantity_dict:
        most_products_order_id = i_d
print(most_products_order_id)

In [ ]:
print(f'La órden con mayor cantidad de productos es la número {most_products_order_id}, con {most_products_in_one_order} productos en total'.)